File: J9Ch/src/S_2_Data_Structure_I/Required_12/_434_Number_of_Islands_II_Hard.py



# Definition for a point.
# class Point:
#     def __init__(self, a=0, b=0):
#         self.x = a
#         self.y = b

class Solution:
    # @param {int} n an integer
    # @param {int} m an integer
    # @param {Pint[]} operators an array of point
    # @return {int[]} an integer array
    def __init__(self):
        self.f = []

    def find(self, x):
        if self.f[x] == x:
            return x
        self.f[x] = self.find(self.f[x])
        return self.f[x]

    def merge(self, x, y):
        if self.f[x] == -1 or self.f[y] == -1:
            return False

        x = self.find(x)
        y = self.find(y)
        if x != y:
            self.f[x] = y
            return True
        else:
            return False

    def inside(self, x, y, n, m):
        return x >= 0 and y >=0 and x < n and y < m

    def numIslands2(self, n, m, operators):
        # Write your code here
        d = [[0,1],[0,-1],[1,0],[-1,0]]
        area = 0;
        cnt = len(operators)
        ret = []
        for i in xrange(n * m): self.f.append(-1)
        for point in operators:
            num = point.x * m + point.y;
            if (self.f[num] == -1):
                area += 1;
                self.f[num] = num

            for k in xrange(4):
                x = point.x + d[k][0];
                y = point.y + d[k][1];

                if self.inside(x, y, n, m):
                    if self.merge(x * m + y, num):
                        area -= 1

            ret.append(area);

        return ret

File: J9Ch/src/S_2_Data_Structure_I/Required_12/_434_Number_of_Islands_II_Hard.cpp


/**
 * Definition for a point.
 * struct Point {
 *     int x;
 *     int y;
 *     Point() : x(0), y(0) {}
 *     Point(int a, int b) : x(a), y(b) {}
 * };
 */
class Solution {
public:
    /**
     * @param n an integer
     * @param m an integer
     * @param operators an array of point
     * @return an integer array
     */
    int find(int x) {
        if (f[x] == x)
            return x;
        return f[x] = find(f[x]);
    }

    bool merge(int x, int y) {
        if (f[x] == -1 || f[y] == -1)
            return false;

        x = find(x);
        y = find(y);
        if (x != y) {
            f[x] = y;
            return true;
        } else
            return false;
    }

    bool inside(int x, int y, int n, int m) {
        return (x >= 0 && y >=0 && x < n && y < m);
    }

    int f[100000];
    vector<int> numIslands2(int n, int m, vector<Point>& operators) {
        // Write your code here
        const int d[4][2] = {0, 1, 0, -1, 1, 0, -1, 0};
        int area = 0;
        int cnt = operators.size();
        vector<int> ret;
        for (int i = 0; i < n * m; ++i) f[i] = -1;
        for (int i = 0; i < cnt; ++i) {
            int point = operators[i].x * m + operators[i].y;
            if (f[point] == -1) {
                area += 1;
                f[point] = point;
            }
            for (int k = 0; k < 4; ++k) {
                int x = operators[i].x + d[k][0];
                int y = operators[i].y + d[k][1];

                if (inside(x, y, n, m)) {
                    if (merge(x * m + y, point))
                        area -= 1;
                }
            }
            ret.push_back(area);
        }
        return ret;
    }
};